src/cargo/util/context/de.rs

```diff
@@ -265,8 +271,14 @@ impl<'gctx> ConfigMapAccess<'gctx> {
             let mut field_key = de.key.clone();
             field_key.push(field);
             for env_key in de.gctx.env_keys() {
-                if env_key.starts_with(field_key.as_env_key()) {
```

Comment:
See https://github.com/rust-lang/cargo/pull/13687#discussion_r1607775037

I wonder if this is worth fixing on separately PR.  Because it's only a problem in my situation

Reply:
Probably. I'll find a time reviewing this.
(it's always a headache for me to read config deserialization code 🤯)

Reply:
Just reviewed it again. Yeah I believe this can pull out into a separate PR.

Also, we improve it a bit to avoid extra allocation and string comparsion:

```
            for env_key in de.gctx.env_keys() {
                let Some(nested_field) = env_key.strip_prefix(field_key.as_env_key()) else {
                    continue;
                };
                // This distinguishes fields that share the same prefix.
                // For example, when env_key is UNSTABLE_GITOXIDE_FETCH
                // and field_key is UNSTABLE_GIT, the field shouldn't be 
                // added because `unstable.gitoxide.fetch` doesn't
                // belong to `unstable.git` struct.
                if nested_field.is_empty() || nested_field.starts_with('_') {
                    fields.insert(KeyKind::Normal(field.to_string()));
                }
            }
```